Readme.md
[![Build Status](https://img.shields.io/travis/Authmaker/authmaker-verify-express/master.svg)](https://travis-ci.org/Authmaker/authmaker-verify-express)
[![Code Climate](https://img.shields.io/codeclimate/github/Authmaker/authmaker-verify-express.svg)](https://codeclimate.com/github/Authmaker/authmaker-verify-express)

[![Version npm](https://img.shields.io/npm/v/authmaker-verify-express.svg)](https://www.npmjs.com/package/authmaker-verify-express)
[![Dependencies](https://img.shields.io/david/Authmaker/authmaker-verify-express.svg)](https://david-dm.org/Authmaker/authmaker-verify-express)
[![npm Downloads](https://img.shields.io/npm/dm/authmaker-verify-express.svg)](https://www.npmjs.com/package/authmaker-verify-express)

# Authmaker Verify Express
This package allows you to use [Authmaker verify](https://www.npmjs.com/package/authmaker-verify) extremely easily in an ExpressJS based Node application.

## Installation
```
npm install --save authmaker-verify-express
```

### Usage
This package currently makes use of the Mongo connection for Authmaker Verify so you need to initialise the database connection before use:
```
var authmakerVerify = require('authmaker-verify');
authmakerVerify.connectMongo(nconf);
```

You need to pass a [nconf](https://github.com/indexzero/nconf) object into the connectMongo call that has access to at least the following parameters:

```
{
    "mongo": {
        "authmaker": {
            "db": "your-db-name",
            "host": "localhost",
            "port": 27017
        }
    }
}
```

you can also optionally include `username` and `password`. Each of these config entries are accessed asynchronously so you can use any of the asynchronous stores for nconf

#### Middlewares
To actually use this package you just need to include middlewares in your ExpressJS app. Here are a few examples of ways you can use authmaker-verify-express. Please note that all of these examples use a simple "success" callback that does nothing but responds with a 200 response code. They have also already imported required modules:

```
var authmakerVerifyExpress = require('authmaker-verify-express');
var express = require('express');
var app = express();

function success(req, res){
    return res.send("Success");
}
```

Requires users with valid, in date access tokens in the request:
```
app.get('/verify': [authmakerVerifyExpress.mongo(), success]);
```

Requires users with valid, in date access tokens with the scope "my_awesome_permission":
```
app.get('/scope': [authmakerVerifyExpress.mongo("my_awesome_permission"), success]);
```

Requires users with valid, in date access tokens with a rate limited scope (suffix `_limit_<num>_<timeframe>`)
```
app.get('/jointrated': [authmakerVerifyExpress.mongoRateLimited("face"), success]);
```

Requires users with valid, in date access token but if they don't have a rate limited scope it uses `face_limit_10_minutes` as a default scope:
```
app.get('/defaultScope': [authmakerVerifyExpress.mongoRateLimited("face", "face_limit_10_minutes"), success]);
```

### API

#### mongoRateLimited - function - returns middleware
```
mongoRateLimited: function(tag, defaultScope)
```

#### mongoRateLimitedDefault - function - returns middleware
```
mongoRateLimitedDefault: function(tag, defaultScope)
```

#### mongo - function - returns middleware
```
mongo: function(tag, options)
```

options.passError (optional) if true passes error via middleware

#### connectMongo - function
```
connectMongo: function(nconf) {
    //initialise the db
    authmakerVerify.connectMongo(nconf);
}
```

#### authmakerVerify - object
If you ever need to access the [authmaker-verify](https://github.com/Authmaker/authmaker-verify) object that is powering authmaker-verify-express to access any lower level apis you can access it directly like this:
```
var authmakerVerifyExpress = require('authmaker-verify-express');
authmakerVerifyExpress.authmakerVerify;
```


History.md

3.0.1 / 2019-06-13
==================

  * bumping dependencies

2.2.0 / 2016-03-31
==================

  * Merge pull request #7 from Authmaker/feature/addingExternalIdentities
  * adding external identities
  * fixing passerror
  * updating to our new coding standards

2.1.0 / 2015-12-18
==================

  * Merge pull request #6 from Authmaker/feature/addUserReq
  * adding req.user

2.0.0 / 2015-12-18
==================

  * Merge pull request #5 from Authmaker/feature/updateAuthmakerVerify
  * upgrading Authmaker Verify and fixing tests

1.2.1 / 2015-12-04
==================

  * added missing options param in mongo export
  * Merge pull request #3 from Authmaker/release/1.2.0

1.2.0 / 2015-12-04
==================

  * Merge pull request #2 from Authmaker/feature/addOptions
  * test fix
  * added options to generateVerify

n.n.n / 2015-12-04
==================

  * Merge pull request #2 from Authmaker/feature/addOptions
  * test fix
  * added options to generateVerify

n.n.n / 2015-12-04
==================

  * Merge pull request #2 from Authmaker/feature/addOptions
  * test fix
  * added options to generateVerify

1.0.4 / 2015-07-02
==================

  * deprecating mongoRateLimitedDefault


test/index.js
const http = require('http');
const nconf = require('nconf');
const verify = require('authmaker-verify');

const express = require('express');

const authmakerVerify = require('authmaker-verify');

let httpServer;

nconf.defaults({
  authmaker: {
    mongo: {
      db: 'authmaker-verify-express-test',
      host: 'localhost',
      port: 27017,
    },
  },
});

before(() => {
  global.app = express();
  // create http server
  httpServer = http.createServer(global.app).listen(56773);
  return authmakerVerify.init(nconf);
});

after(() => {
  httpServer.close();
  return verify.getConnection().then(connection => connection.close());
});


test/routes/mongo.js
var expect = require('chai').expect;
var request = require('supertest');
var autoroute = require('express-autoroute');
var Q = require('q');

var fixture = rootRequire('test/fixtures/users');

function promiseRequest(params) {
  var deferred = Q.defer();

  var req = request(global.app)
    .get(params.route);

  if (params.access_token) {
    req.set('authorization', 'bearer: ' + params.access_token);
  }

  req.expect(200)
    .end(function(err) {
      if (err) {
        return deferred.reject(err);
      }

      deferred.resolve();
    });

  return deferred.promise;
}

function manyRequests(params, number, done) {
  var promises = [];

  for (var i = 0; i < number; i++) {
    promises.push(promiseRequest(params));
  }

  Q.all(promises).then(function() {
    done();
  }, function(err) {
    done(err);
  });
}

describe('mongo verify functions', function() {

  before(function() {
    autoroute(global.app, {
      throwErrors: true,
      routeFile: rootPath('test/fixtures/routes/mongo.js'),
    });
  });

  beforeEach(function() {
    return fixture.init();
  });

  afterEach(function() {
    fixture.reset();
  });

  it('should success every time with no verify', function(done) {
    request(global.app)
      .get('/noverify')
      .expect(200, done);
  });

  describe('verify only', function() {
    it('should fail when there is no access token', function(done) {
      request(global.app)
        .get('/verify')
        .expect(401)
        .expect(function(res) {
          expect(res).to.have.property('text', 'Not Authorized: No Access token provided');
        })
        .end(done);
    });

    it('should fail when there is an access token but it doesn\'t match the database', function(done) {
      request(global.app)
        .get('/verify')
        .set('authorization', 'Bearer: this_doesnt_really_exist')
        .expect(401)
        .expect(function(res) {
          expect(res).to.have.property('text', 'Not Authorized: session not found with that access token');
        })
        .end(done);
    });

    it('should succeed when there is a valid access token', function(done) {
      request(global.app)
        .get('/verify')
        .set('authorization', 'bearer: valid_access_token_1')
        .expect(200)
        .expect(function(res) {
          expect(res).to.have.property('text', 'Success');
        })
        .end(done);
    });

    it('it should fail if there is a valid, expired access token', function(done) {
      request(global.app)
        .get('/verify')
        .set('authorization', 'bearer: expired_access_token')
        .expect(401)
        .expect(function(res) {
          expect(res).to.have.property('text', 'Not Authorized: session has expired');
        })
        .end(done);
    });

    it('should fail if the endpoint needs a scope permission but the session doesn\'t have any', function(done) {
      request(global.app)
        .get('/scope')
        .set('authorization', 'bearer: valid_access_token_1')
        .expect(401)
        .expect(function(res) {
          expect(res).to.have.property('text', 'Not Authorized: No scope associated with face_permissions');
        })
        .end(done);
    });

    it('should fail if the endpoint needs a scope permission but the session has the wrong one', function(done) {
      request(global.app)
        .get('/scope')
        .set('authorization', 'bearer: valid_rate_limit_5_days')
        .expect(401)
        .expect(function(res) {
          expect(res).to.have.property('text', 'Not Authorized: No scope associated with face_permissions');
        })
        .end(done);
    });

    it('should succeed if the endpoint needs a scope permission and it is on the session', function(done) {
      request(global.app)
        .get('/scope')
        .set('authorization', 'bearer: valid_face_permission')
        .expect(200)
        .expect(function(res) {
          expect(res).to.have.property('text', 'Success');
        })
        .end(done);
    });
  });

  describe('rate limited', function() {
    it('should fail when there is no access token', function(done) {
      request(global.app)
        .get('/jointrated')
        .expect(401)
        .expect(function(res) {
          expect(res).to.have.property('text', 'No Access token provided');
        })
        .end(done);
    });

    it('should fail when there is an access token but it doesn\'t match the database', function(done) {
      request(global.app)
        .get('/jointrated')
        .set('authorization', 'Bearer: this_doesnt_really_exist')
        .expect(401)
        .expect(function(res) {
          expect(res).to.have.property('text', 'Not Authorized: session not found with that access token');
        })
        .end(done);
    });

    it('it should fail if there is a valid, expired access token', function(done) {
      request(global.app)
        .get('/jointrated')
        .set('authorization', 'bearer: expired_access_token')
        .expect(401)
        .expect(function(res) {
          expect(res).to.have.property('text', 'Not Authorized: session has expired');
        })
        .end(done);
    });

    it('should fail if there is no scope and no default', function(done) {
      request(global.app)
        .get('/jointrated')
        .set('authorization', 'bearer: expired_access_token')
        .expect(401)
        .expect(function(res) {
          expect(res).to.have.property('text', 'Not Authorized: session has expired');
        })
        .end(done);
    });

    it('should suceed if there is a default scope', function(done) {
      request(global.app)
        .get('/defaultScope')
        .set('authorization', 'bearer: valid_access_token_1')
        .expect(200)
        .end(done);
    });

    it('should allow 5 requests for a scope that ends with _limit_5_days', function(done) {
      manyRequests({
        route: '/jointrated',
        access_token: 'valid_rate_limit_5_days'
      }, 5, done);
    });

    it('should fail with 6 requests for a scope that ends with _limit_5_days', function(done) {
      manyRequests({
        route: '/jointrated',
        access_token: 'valid_rate_limit_5_days'
      }, 5, function(err) {
        if (err) {
          return done(err);
        }

        request(global.app)
          .get('/jointrated')
          .set('authorization', 'bearer: valid_rate_limit_5_days')
          .expect(429)
          .expect(function(res) {
            expect(res).to.have.property('text', 'Too Many Requests: Rate limit exceeded.');
          })
          .end(done);
      });
    });

    it('should allow 10 requests for a scope that ends with _limit_5_seconds where there is a 1 second gap after the first 5 requests', function(done) {

      this.timeout(6000);
      manyRequests({
        route: '/jointrated',
        access_token: 'valid_rate_limit_5_second'
      }, 5, function(err) {
        if (err) {
          return done(err);
        }

        setTimeout(function() {
          manyRequests({
            route: '/jointrated',
            access_token: 'valid_rate_limit_5_second'
          }, 5, done);
        }, 1000);
      });
    });

    it('should fail with 11 requests for a scope that ends with _limit_5_seconds where there is a 1 second gap after the first 5 requests', function(done) {
      manyRequests({
        route: '/jointrated',
        access_token: 'valid_rate_limit_5_second'
      }, 5, function(err) {
        if (err) {
          return done(err);
        }

        setTimeout(function() {
          manyRequests({
            route: '/jointrated',
            access_token: 'valid_rate_limit_5_second'
          }, 5, function(err) {
            if (err) {
              return done(err);
            }

            request(global.app)
              .get('/jointrated')
              .set('authorization', 'bearer: valid_rate_limit_5_second')
              .expect(429)
              .expect(function(res) {
                expect(res).to.have.property('text', 'Too Many Requests: Rate limit exceeded.');
              })
              .end(done);
          });
        }, 1000);
      });
    });

    it('should allow 2 users to send 5 requests each for a scope that ends with _limit_5_days', function(done) {
      manyRequests({
        route: '/jointrated',
        access_token: 'valid_rate_limit_5_days'
      }, 5, function(err) {
        if (err) {
          return done(err);
        }

        manyRequests({
          route: '/jointrated',
          access_token: 'valid_rate_limit_5_days_user2'
        }, 5, done);
      });
    });
  });
});


test/fixtures/users.js
const authmakerVerify = require('authmaker-verify');
const moment = require('moment');
const mongoose = require('mongoose');
const Q = require('q');

const sessionsToCreate = [{
  accessToken: 'valid_access_token_1',
  expiry: moment().add(1, 'hours').toDate(),
  userId: mongoose.Types.ObjectId(),
}, {
  accessToken: 'valid_face_permission',
  expiry: moment().add(1, 'hours').toDate(),
  scopes: ['face_permissions'],
  userId: mongoose.Types.ObjectId(),
}, {
  accessToken: 'expired_access_token',
  expiry: moment().subtract(1, 'seconds').toDate(),
  userId: mongoose.Types.ObjectId(),
}, {
  accessToken: 'valid_rate_limit_5_days',
  expiry: moment().add(1, 'hours').toDate(),
  scopes: ['face_limit_5_days'],
  userId: mongoose.Types.ObjectId(),
}, {
  accessToken: 'valid_rate_limit_5_days_user2',
  expiry: moment().add(1, 'hours').toDate(),
  scopes: ['face_limit_5_days'],
  userId: mongoose.Types.ObjectId(),
}, {
  accessToken: 'valid_rate_limit_5_second',
  expiry: moment().add(1, 'hours').toDate(),
  scopes: ['face_limit_5_seconds'],
  userId: mongoose.Types.ObjectId(),
}];

function init() {
  return authmakerVerify.models.oauthSession.create(sessionsToCreate);
}

function reset() {
  // only allow this in test
  if (process.env.NODE_ENV === 'test') {
    return authmakerVerify.getConnection().then(function (connection) {
      const collections = connection.collections;

      const promises = Object.keys(collections).map(function (collection) {
        return Q.ninvoke(collections[collection], 'remove');
      });

      return Q.all(promises);
    });
  }
  const errorMessage = 'Excuse me kind sir, but may I enquire as to why you are currently running reset() in a non test environment? I do propose that it is a beastly thing to do and kindly ask you to refrain from this course of action. Sincerely yours, The Computer.';
  console.log(errorMessage);
  console.error(errorMessage);
  throw new Error(errorMessage);
}

module.exports = {
  init,
  reset,
};


index.js
const authmakerVerify = require('authmaker-verify');
const winston = require('winston');
const Q = require('q');

// remove it so to add it with my settings
try {
  winston.remove(winston.transports.Console);
} catch (e) { // do nothing
}

const winstonOptions = {
  colorize: true,
  timestamp: true,
  handleExceptions: true,
  prettyPrint: true,
};

if (process.env.LOG_LEVEL) {
  winstonOptions.level = process.env.LOG_LEVEL;
} else if (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test') {
  winstonOptions.level = 'debug';
} else {
  winstonOptions.level = 'info';
}

if (!process.env.NO_LOG) {
  winston.add(winston.transports.Console, winstonOptions);
}

function generateRateLimit(tag, defaultScope) {
  return function rateLimit(req, res, next) {
    if (!req.headers.authorization) {
      return res.status(401).send('No Access token provided');
    }

    // take the accessToken as the last delimited entry in authorization
    const accessToken = req.headers.authorization.split(/\s+/).pop();

    // verify the access-token
    return authmakerVerify.mongoRateLimited(accessToken, tag, defaultScope)

      .then((oauthSession) => {
        req.oauthSession = oauthSession;

        // add the user to the req.user
        return authmakerVerify.models.user.findOne({
          _id: oauthSession.userId,
        }).exec().then((user) => {
          req.user = user;
        });
      })

      .then(() => {
        next();
      })

      .then(null, (err) => {
        winston.error('Error while authorizing session', {
          error: err.message,
          stask: err.stack,
          authorisation: req.headers.authorization,
        });

        if (err.message.indexOf('Not Authorized') >= 0) {
          res.status(401);
        } else if (err.message.indexOf('Too Many Requests') >= 0) {
          res.status(429);
          return res.send('Too Many Requests: Rate limit exceeded.');
        } else {
          res.status(500);
        }

        return res.send(err.message);
      });
  };
}

function generateVerify(tag, options) {
  return function (req, res, next) {
    // verify the access-token
    return Q.fcall(() => {
      if (!req.headers.authorization) {
        throw new Error('Not Authorized: No Access token provided');
      }

      // take the accessToken as the last delimited entry in authorization
      const accessToken = req.headers.authorization.split(/\s+/).pop();

      return authmakerVerify.mongo(accessToken, tag);
    })

      .then((oauthSession) => {
        req.oauthSession = oauthSession;

        // add the user to the req.user
        return authmakerVerify.models.user.findOne({
          _id: oauthSession.userId,
        }).exec().then((user) => {
          req.user = user;
        });
      })

      .then(() => {
        next();
      })

      .then(null, (err) => {
        winston.error('Error while authorizing session', {
          error: err.message,
          stask: err.stack,
          authorisation: req.headers.authorization,
        });

        if (options && options.passError) {
          return next(err);
        }

        if (err.message.indexOf('Not Authorized') >= 0) {
          res.status(401);
        } else {
          res.status(500);
        }

        return res.send(err.message);
      });
  };
}

function getExternalIdentities(req, res, next) {
  if (!req.user) {
    return next('req.user not defined - must use mongo(), mongoRateLimited() or mongoRateLimitedDefault() before this middleware');
  }

  return authmakerVerify.models.externalIdentity.find({
    _id: req.user.externalIdentities,
  }).exec().then((externalIdentities) => {
    req.externalIdentities = externalIdentities;

    next();
  }).then(null, (err) => {
    next(err);
  });
}

module.exports = {
  mongoRateLimited(tag, defaultScope) {
    return generateRateLimit(tag, defaultScope);
  },

  mongoRateLimitedDefault(tag, defaultScope) {
    console.warn('This function is deprecated, just use mongoRateLimited(tag, defaultScope) instead');
    return generateRateLimit(tag, defaultScope);
  },

  mongo(tag, options) {
    return generateVerify(tag, options);
  },

  externalIdentities: getExternalIdentities,

  init(nconf) {
    // initialise the db
    return authmakerVerify.init(nconf);
  },

  models: authmakerVerify.models,

  authmakerVerify,
};
